vip_hci/negfc/simplex_fmerit.py

```diff
@@ -116,60 +116,91 @@ def get_values_optimize(cube, angs, ncomp, annulus_width, aperture_radius,
-    posy = r_guess * np.sin(np.deg2rad(theta_guess)) + centy_fr
-    posx = r_guess * np.cos(np.deg2rad(theta_guess)) + centx_fr
-    halfw = max(aperture_radius, annulus_width/2.)
-
-    # Checking annulus/aperture sizes. Assuming square frames
-    msg = 'The annulus and/or the circular aperture used by the NegFC falls '
-    msg += 'outside the FOV. Try increasing the size of your frames or '
-    msg += 'decreasing the annulus or aperture size'
-    if r_guess>centx_fr-halfw or r_guess<=halfw:
-        raise RuntimeError(msg)
-        
-    pca_res = pca_annulus(cube, angs, ncomp, annulus_width, r_guess, cube_ref,
-                          svd_mode, scaling, collapse=collapse)
-    indices = circle(posy, posx, radius=aperture_radius)
-    yy, xx = indices
-
-    if collapse is None:
-        values = pca_res[:, yy, xx].ravel()
-    else:
-        values = pca_res[yy, xx].ravel()
+    if cube.ndim==3:
+        centy_fr, centx_fr = frame_center(cube[0])
 
-    if debug and collapse is not None:
-        return values, pca_res
-    else:
-        return values
+        posy = r_guess * np.sin(np.deg2rad(theta_guess)) + centy_fr
+        posx = r_guess * np.cos(np.deg2rad(theta_guess)) + centx_fr
+        halfw = max(aperture_radius, annulus_width/2.)
+
+        # Checking annulus/aperture sizes. Assuming square frames
+        msg = 'The annulus and/or the circular aperture used by the NegFC falls '
+        msg += 'outside the FOV. Try increasing the size of your frames or '
+        msg += 'decreasing the annulus or aperture size'
+        if r_guess>centx_fr-halfw or r_guess<=halfw:
+            raise RuntimeError(msg)
+
+        pca_res = pca_annulus(cube, angs, ncomp, annulus_width, r_guess, cube_ref,
+                            svd_mode, scaling, collapse=collapse)
+        indices = circle(posy, posx, radius=aperture_radius)
+        yy, xx = indices
+
+        if collapse is None:
+            values = pca_res[:, yy, xx].ravel()
+        else:
+            values = pca_res[yy, xx].ravel()
+
+        if debug and collapse is not None:
+            return values, pca_res
+        else:
+            return values
+
+    if cube.ndim==4:
+        centy_fr, centx_fr = frame_center(cube[0,0])
+        posy = r_guess * np.sin(np.deg2rad(theta_guess)) + centy_fr
+        posx = r_guess * np.cos(np.deg2rad(theta_guess)) + centx_fr
+        halfw = max(aperture_radius, annulus_width/2.)
+
+        # Checking annulus/aperture sizes. Assuming square frames
+        msg = 'The annulus and/or the circular aperture used by the NegFC falls '
+        msg += 'outside the FOV. Try increasing the size of your frames or '
+        msg += 'decreasing the annulus or aperture size'
+        if r_guess>centx_fr-halfw or r_guess<=halfw:
+            raise RuntimeError(msg)
+
+        pca_res = np.zeros_like(cube[0])
+
+        for i in range(cube[0].shape):
+            pca_res[i] = pca_annulus(cube[i], angs, ncomp, annulus_width, r_guess, cube_ref,
```

Comment:
@arainot In this line we are slicing the 4D cube, which means that we're passing a 3D cube to ``pca_annulus``. Isn't it? Then I come back to my previous question, why we need ``pca_annulus`` to handle 4D cubes?

Reply:
@carlgogo Yes indeed that's weird, I'll take that off and put it back in a PR if I encounter the need for it later.

Reply:
@arainot So did you confirm that with the old version of ``pca_annulus`` (which only handles 3D arrays) you could call ``get_values_optimize`` with a 4D cube successfully? If yes, then I don't think I have other comments

Reply:
@carlgogo Yes it does work, because we call pca_annulus in 3D from the 4D case. I have just tested if we could do directly pca_annulus in 4D and it does not seem to work. There is an issue with the matrices going from 3D to 4D. Would you like me to remove it or do you want to have a look at it and maybe find a fix? 
I was not able to find the source.